controller/raft/fsm.go

```diff
@@ -0,0 +1,146 @@
+package raft
+
+import (
+	"github.com/hashicorp/raft"
+	"github.com/openziti/fabric/controller/command"
+	"github.com/openziti/fabric/controller/db"
+	"github.com/openziti/foundation/util/concurrenz"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"go.etcd.io/bbolt"
+	"io"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"sync/atomic"
+)
+
+func NewFsm(dataDir string, decoders command.Decoders) *BoltDbFsm {
+	return &BoltDbFsm{
+		dataDir:  dataDir,
+		decoders: decoders,
+		dbPath:   dataDir + "ctrl.db",
+	}
+}
+
+type BoltDbFsm struct {
+	db          *db.Db
+	dataDir     string
+	dbPath      string
+	decoders    command.Decoders
+	env         atomic.Value
+	initialized concurrenz.AtomicBoolean
+}
+
+func (self *BoltDbFsm) Init() error {
+	var err error
+	self.db, err = db.Open(self.dbPath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (self *BoltDbFsm) RaftInitialized() {
+	self.initialized.Set(true)
+}
+
+func (self *BoltDbFsm) GetDb() *db.Db {
+	return self.db
+}
+
+func (self *BoltDbFsm) Apply(log *raft.Log) interface{} {
+	if log.Type == raft.LogCommand {
+		if len(log.Data) >= 4 {
+			cmd, err := self.decoders.Decode(log.Data)
+			if err != nil {
+				logrus.WithError(err).Error("failed to create command")
+				return err
+			}
+
+			logrus.Infof("apply log with type=%v", reflect.TypeOf(cmd))
```

Comment:
can use `%T` here as an alternative to `TypeOf`

Reply:
Fixed